PMM-11167: Improve helm documentation (#984)
percona/pmm-doc: docs/setting-up/server/helm.md

!!! note alert alert-primary ""
    Helm v3 is needed to run the following steps.

Please refer to the [Kubernetes Supported versions](https://kubernetes.io/releases/version-skew-policy/#supported-versions) and [Helm Version Support Policy](https://helm.sh/docs/topics/version_skew/) to find out supported versions.

PMM should be platform agnostic, but so far, it requires escalated privileges inside of a container. `root` user inside the PMM container is needed. Thus PMM would not work for Kubernetes Platforms such as OpenShift or others that have hardened Security Context Constraints, for example:

- https://docs.openshift.com/container-platform/latest/security/container_security/security-platform.html#security-deployment-sccs_security-platform
- https://docs.openshift.com/container-platform/latest/authentication/managing-security-context-constraints.html

Kubernetes platforms offer a different set of capabilities. To use PMM in production, you would need backups and, thus storage driver that supports snapshots. Please refer to your Kubernetes and Cloud provider for storage capabilities.

> ## Locality and Availability

Author: yes, on what nodes, regions and etc

> ## Locality and Availability

Reviewer: The word Locality doesn't seem proper..

It is not recommended to run the PMM monitoring server on the same system together with monitored database clusters and services.

Please ensure proper locality either by physically separating workloads in Kubernetes clusters or running separate Kubernetes clusters for the databases and monitoring workloads.

Physical separation of workloads can be done with proper Kubernetes nodes configuration, affinity rules, label selections and etc.

> Please also ensure the Kubernetes cluster has [high availability](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/ha-topology/) so that in a case of a node failure, the monitoring service will be running and capturing needed data.

Reviewer: ```suggestion
Also, ensure that the Kubernetes cluster has [high availability](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/ha-topology/) so that in case of a node failure, the monitoring service will be running and capturing the required data.
```

> PMM, when correctly set up, would be highly available together with Kubernetes proper HA. Please ensure that your network and storage support high availability. In case of Kubernetes node failure - PMM will be rescheduled and available on any other node in a Kubernetes cluster.

Reviewer: ```suggestion
When configured correctly, PMM, along with Kubernetes' high availability, would be highly available. Ensure that your network and storage support high availability. In the event of a Kubernetes node failure, PMM will be scheduled on any other node in the Kubernetes cluster.
```

## Use Helm to install PMM server on Kubernetes clusters

!!! note alert alert-primary "Availability"
    This feature is available starting with PMM 2.29.0.

!!! summary alert alert-info "Summary"
    - Setup PMM admin password
    - Install
    - Configuration parameters
    - PMM environment variables
    - PMM SSL certificates
    - Backup
    - Upgrade
    - Restore
    - Uninstall

---

### Setup PMM admin password

Create Kubernetes secret with PMM admin password:
```sh
cat <<EOF | kubectl create -f -
apiVersion: v1
kind: Secret
metadata:
  name: pmm-secret
type: Opaque
data:
# base64 encoded password
# encode some password: `echo -n "admin" | base64`
  PMM_ADMIN_PASSWORD: YWRtaW4=
EOF
```

To get admin password execute:

```sh
kubectl get secret pmm-secret -o jsonpath='{.data.PMM_ADMIN_PASSWORD}' | base64 --decode
```

### Install

To install the chart with the release name `pmm`:

```sh
helm repo add percona https://percona.github.io/percona-helm-charts/
helm install pmm percona/pmm --set secret.create=false --set secret.name=pmm-secret
```

> The command deploys PMM on the Kubernetes cluster in the default configuration and specified secret. The [Parameters](#parameters) section lists the parameters that can be configured during installation.

Reviewer: the term **and specified secret** is not clear

Author: `with secrets that was created earlier`?

!!! note alert alert-primary ""
    Helm v3 is needed to run the following steps.

Refer to [Kubernetes Supported versions](https://kubernetes.io/releases/version-skew-policy/#supported-versions) and [Helm Version Support Policy](https://helm.sh/docs/topics/version_skew/) to find the supported versions.

> PMM should be platform-agnostic, but it requires escalated privileges inside a container. It is necessary to have a `root` user inside the PMM container. Thus, PMM would not work for Kubernetes Platforms such as OpenShift or others that have hardened Security Context Constraints, for example:

Reviewer 2: This can be run on openshift as well but with required SCC.

Author: yes, I think that what I wrote there ? that SCC is limiting it.

- [Managing security context constraints](https://docs.openshift.com/container-platform/latest/authentication/managing-security-context-constraints.html)

Kubernetes platforms offer a different set of capabilities. To use PMM in production, you would need backups and, thus storage driver that supports snapshots. Consult your provider for Kubernetes and Cloud storage capabilities.

## Locality and Availability

You should not run the PMM monitoring server along with the monitored database clusters and services on the same system.

Please ensure proper locality either by physically separating workloads in Kubernetes clusters or running separate Kubernetes clusters for the databases and monitoring workloads.

You can physically separate workloads by properly configuring Kubernetes nodes, affinity rules, label selections, etc.

Also, ensure that the Kubernetes cluster has [high availability](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/ha-topology/) so that in case of a node failure, the monitoring service will be running and capturing the required data.

> When configured correctly, PMM, along with Kubernetes' high availability, would be highly available. Ensure that your network and storage support high availability. In the event of a Kubernetes node failure, PMM will be scheduled on any other node in the Kubernetes cluster.

Reviewer 2: I am extremely skpetical of this. Mentioning PMM as HA now is not right.

Reviewer 3: The redaction of this phrase can be improved: "When configured correctly, PMM, along with Kubernetes' high availability, would be highly available."

Now, what is "configured correctly" ? and what is the definition of HA here? Seamless operation? resilient? i agree with @cshiv : not sure if is okay to claim that PMM is HA just because it runs over a k8s cluster

Author: CC @dygksy.

I don't mind to remove HA. StatefulSet and PVC provide you HA for the Node failure.

Configured properly - is a hell of another doc on how to do it (mainly how to allocate HA CSI drive, and HA network DNS).

Author: also that HA with drives and net is very cloud dependable, so documenting it for each could is quite a task.

> 

Reviewer 4: @denisok why dont we use [this paragraph](https://docs.google.com/document/d/1A3t8pYciPB8xyeD_iZhl2mNZtI-H704xp5yYOSU0k9U/edit#bookmark=id.gq45i8o4hzxw)

Author: I would vote to remove this HA section at all to not confuse ppl.

## Use Helm to install PMM server on Kubernetes clusters

!!! note alert alert-primary "Availability"
    This feature is available starting with PMM 2.29.0.

!!! summary alert alert-info "Summary"
    - Setup PMM admin password
    - Install
    - Configuration parameters
    - PMM environment variables
    - PMM SSL certificates
    - Backup
    - Upgrade
    - Restore
    - Uninstall

---

### Setup PMM admin password

Create Kubernetes secret with PMM admin password:

> ```sh

Reviewer 5: If you wanted a hard-copy to review afterwards then:
```sh
cat <<EOF | tee /path/to/saved/output | kubectl create -f -
```

> ```sh

Author: depends how ppl use it. I suppose as part of a pipeline, so they need only yaml part of it.

I use bash style mostly to have it covered by tests as well:
https://github.com/percona/pmm-doc/blob/main/.github/workflows/helm-tests.yml

> ```sh

Reviewer 5: Are you trying to improve the docs and make them easier?

```sh
helm show values percona/pmm > values.yaml

#change needed parameters in values.yaml

> yq -i e '.secret.create |= false' values.yaml

Reviewer 5: This is not a common package and I don't see mention of what it is anywhere, nor how to get it

Author: but that is not a point, just to demonstrate that something needs to be changed. It also to enable testing: https://github.com/percona/pmm-doc/blob/main/.github/workflows/helm-tests.yml

    dhparam.pem: <content>
```

> Another approach to set up TLS certificates is to use Ingress controller, see [TLS](https://kubernetes.io/docs/concepts/services-networking/ingress/#tls). PMM helm chart supports Ingress; please see [PMM network configuration](https://github.com/percona/percona-helm-charts/tree/main/charts/pmm#pmm-network-configuration).

Reviewer: ```suggestion
Another approach to set up TLS certificates is to use the Ingress controller, see [TLS](https://kubernetes.io/docs/concepts/services-networking/ingress/#tls). PMM helm chart supports Ingress. See [PMM network configuration](https://github.com/percona/percona-helm-charts/tree/main/charts/pmm#pmm-network-configuration).
```

## Backup

PMM helm chart uses [PersistentVolume and PersistentVolumeClaim](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) to allocate storage in the Kubernetes cluster.

Volumes could be pre-provisioned and dynamic. PMM chart supports both and exposes it through [PMM storage configuration](https://github.com/percona/percona-helm-charts/tree/main/charts/pmm#pmm-storage-configuration).

Backups for the PMM server currently support only storage layer backups and thus require [StorageClass](https://kubernetes.io/docs/concepts/storage/storage-classes/) and [VolumeSnapshotClass](https://kubernetes.io/docs/concepts/storage/volume-snapshot-classes/).

> You could validate proper configuration by these commands:

Reviewer: ```suggestion
Validate the correct configuration by using these commands:
```

```sh
kubectl get sc
kubectl get volumesnapshotclass
```

!!! note alert alert-primary "Storage"

>     Storage configuration is Hardware and Cloud specific. There could be additional costs associated with Volume Snapshots. Please check the documentation for your Cloud or for your Kubernetes cluster.

Reviewer: ```suggestion
    Storage configuration is Hardware and Cloud specific. There could be additional costs associated with Volume Snapshots. Check the documentation for your Cloud or for your Kubernetes cluster.
```

> It is recommended to stop the PMM server before taking the [VolumeSnapshot](https://kubernetes.io/docs/concepts/storage/volume-snapshots/). Here we will stop PMM (scale to 0 pods), take a snapshot, wait for unlit snapshot completes, and start PMM server (scale to 1 pod):

Reviewer: ```suggestion
Before taking a [VolumeSnapshot], stop the PMM server (https://kubernetes.io/docs/concepts/storage/volume-snapshots/). In this step, we will stop PMM (scale to 0 pods), take a snapshot, wait until the snapshot completes, then start PMM server (scale to 1 pod):
```

```sh
kubectl scale statefulset pmm --replicas=0
kubectl wait --for=jsonpath='{.status.replicas}'=0 statefulset pmm

cat <<EOF | kubectl create -f -
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshot
metadata:
  name: before-v2.34.0-upgrade
spec:
  volumeSnapshotClassName: csi-hostpath-snapclass
  source:
    persistentVolumeClaimName: pmm-storage-pmm-0
EOF

kubectl wait --for=jsonpath='{.status.readyToUse}'=true VolumeSnapshot/before-v2.34.0-upgrade
kubectl scale statefulset pmm --replicas=1
```

Output:
```
statefulset.apps/pmm scaled
statefulset.apps/pmm condition met
volumesnapshot.snapshot.storage.k8s.io/before-v2.34.0-upgrade created
volumesnapshot.snapshot.storage.k8s.io/before-v2.34.0-upgrade condition met
statefulset.apps/pmm scaled
```

!!! note alert alert-primary "PMM scale"

>     Only 1 replica set is currently supported.

Reviewer: ```suggestion
    Only one replica set is currently supported.
```

> You can view available snapshots by executing this command:

Reviewer: ```suggestion
You can view available snapshots by executing the following command:
```

Backups for the PMM server currently support only storage layer backups and thus require [StorageClass](https://kubernetes.io/docs/concepts/storage/storage-classes/) and [VolumeSnapshotClass](https://kubernetes.io/docs/concepts/storage/volume-snapshot-classes/).

Validate the correct configuration by using these commands:
```sh
kubectl get sc
kubectl get volumesnapshotclass
```

!!! note alert alert-primary "Storage"
    Storage configuration is Hardware and Cloud specific. There could be additional costs associated with Volume Snapshots. Check the documentation for your Cloud or for your Kubernetes cluster.

Before taking a [VolumeSnapshot](https://kubernetes.io/docs/concepts/storage/volume-snapshots/), stop the PMM server. In this step, we will stop PMM (scale to 0 pods), take a snapshot, wait until the snapshot completes, then start PMM server (scale to 1 pod):
```sh
kubectl scale statefulset pmm --replicas=0
kubectl wait --for=jsonpath='{.status.replicas}'=0 statefulset pmm

cat <<EOF | kubectl create -f -
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshot
metadata:
  name: before-v2.34.0-upgrade
  labels:

>     app.kubernetes.io/name: "pmm"

Reviewer 5: Purely on a standardisation basis... but, why is it double-quoted and nothing else is and this does not contain anything that would need it to be quoted either?

Author: copy-paste from k8s output :)

Reviewer 5: Always review what you c&p :)

This will check updates in the repo and upgrade deployment if the updates are available.

## Restore

> The version of the PMM server should match or be greater than the version in a snapshot. To restore from the snapshot, you would need to delete the old deployment first:

Reviewer: ```suggestion
The version of the PMM server should be greater than or equal to the version in a snapshot. To restore from the snapshot, delete the old deployment first:
```

```sh
helm uninstall pmm
```

And then use snapshot configuration to start the PMM server again with the correct version and correct storage configuration:
```sh
helm install pmm --set image.tag="2.34.0" --set storage.name="pmm-storage-old" --set storage.dataSource.name="before-v2.34.0-upgrade" --set storage.dataSource.kind="VolumeSnapshot" --set storage.dataSource.apiGroup="snapshot.storage.k8s.io" --set secret.create=false --set secret.name=pmm-secret percona/pmm
```

> Here we created a new `pmm-storage-old` PVC with data from the snapshot. So there are a couple of PV and PVCs available in a cluster.

Reviewer: ```suggestion
Here, we created a new `pmm-storage-old` PVC with data from the snapshot. So, there are a couple of PV and PVCs available in a cluster.
```

Reviewer: Give the full form of PV, PVCs at the first instance

This will check updates in the repo and upgrade deployment if the updates are available.

> 

Author: We can, but I think it should be in General section: https://github.com/percona/pmm-doc/blob/main/docs/setting-up/server/index.md

As it applies to all of the install methods.

This command takes a release name and uninstalls the release.

It removes all resources associated with the last release of the chart as well as the release history.

Helm will not delete PVC, PV, and any snapshots. Those need to be deleted manually.

> Also, delete PMM `Secret` if no longer needed:

Reviewer: ```suggestion
Also, delete PMM `Secret` if no longer required:
```
